data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwardingProcessorDecorator.java

```diff
@@ -79,7 +80,9 @@ private PeerForwardingProcessorDecorator(final PeerForwarder peerForwarder, fina
 
     @Override
     public Collection<Record<Event>> execute(final Collection<Record<Event>> records) {
-        final Collection<Record<Event>> recordsToProcessOnLocalPeer = peerForwarder.forwardRecords(records);
+        final Collection<Record<Event>> recordsToProcess = ((RequiresPeerForwarding)innerProcessor).applicableEventsForPeerForwarding(records);
```

Comment:
At the end of this method, we return:

```
return innerProcessor.execute(recordsToProcessLocally);
```

The collection returned by the processor will be the input into the next processor. With this PR, the returned collection is going to be inaccurate when `aggregate_when` is present. The pipeline author still wants to process those events in downstream processors. 

We need something like the following. I'll use `+` for set union and `-` for set difference. This should convey the idea, but these are not valid set operators in Java, so you'll need a little modification. Also, `Collection` is not necessarily a `Set`.

```
return innerProcessor.execute(recordsToProcessLocally) + (records - recordsToProcess);
```

We'll definitely need a unit test here.

It would be ideal to also have a core integration test.